crowbar_framework/app/models/service_object.rb

```diff
@@ -1079,18 +1150,37 @@ def self.proposal_to_role(proposal, bc_name)
   end
 
   #
-  # After validation, this is where the role is applied to the system
-  # The old instance (if one exists) is compared with the new instance.
-  # roles are removed and delete roles are added (if they exist) for nodes leaving roles
-  # roles are added for nodes joining roles.
-  # Calls chef-client on nodes
+  # After validation, this is where the role is applied to the system The old
+  # instance (if one exists) is compared with the new instance.  roles are
+  # removed and delete roles are added (if they exist) for nodes leaving roles
+  # roles are added for nodes joining roles.  Calls chef-client on nodes
   #
-  # This function can be overriden to define a barclamp specific operation.
-  # A call is provided that receives the role and all string names of the nodes before the chef-client call
+  # This function can be overriden to define a barclamp specific operation.  A
+  # call is provided that receives the role and all string names of the nodes
+  # before the chef-client call
   #
+  # The in_queue signifies if apply_role was called from deployment queue's
+  # process_queue, and prevents recursion.
   def apply_role(role, inst, in_queue)
     @logger.debug "apply_role(#{role.name}, #{inst}, #{in_queue})"
 
+    # Part I: Looking up data & checks
+    #
+    # we look up the role in the database (if there is one), the new one
+    # is passed in as the role param.
+    #
+    # From both, we need 'elements', i.e. role -> nodes map and element_order
+    # -> an ordered list of roles, telling us in which order they should be applied.
+    # FIXME: does this give us dependency info within a barclamp?
+    #
+    # Any of the new role's elements can contain clusters, so we need to expand
+    # them to individual nodes. We store them in 'elements_expanded'.
+    # FIXME: would this work if someone reassigned/removed a node from a cluster
+    # in the old role?
```

Comment:
I think it should work, with magic in barclamp-pacemaker updating the roles when a cluster is updated.